### DISCUSSION TITLE
How to create a tool that imports config code shared with apps?

### DISCUSSION THREAD

**Discussion by jtlapp** [nrwl/nx#14731] [General]
I'm struggling to create a tool that is not part of an app but that shares code with apps.

Specifically, I'm using the [kysely query builder](https://github.com/koskimas/kysely), which instead of providing a CLI for database migrations, requires that we write our own migration tool using the kysely library.

I intend for my project to eventually be installed by people with minimal technical skills, so I'm creating a series of configuration classes that validate the data and provide friendly error messages for invalid configuration. One of those classes is `DatabaseConfig` in `libs/config`, which I'd like to use from both the migration tool and the applications.

But I can't seem to create the tool without running into trouble. Here's what I've tried:

- A tool in the `tools/` directory, but it won't import from `@myrepo/config`.
- A library in the `libs/` directory, but `ts-node` complains about `import` statements not being allowed for the given `tsconfig`.
- An nx-plugin with a generator, but it won't import from `@myrepo/config`.

I have not yet tried putting the tool in the `apps/` directory, but maybe that's where it needs to go?

What are my options?

**Comment by jtlapp** [comment]
Well, I am able to make this work as an app in `apps/`. I guess I do eventually need a tool that admins can run to update the installation, and this app could evolve into that tool, so maybe this is where it belongs after all.

**Comment by bcabanes** [comment]
Hi @jtlapp 👋 

The issue is that there is a `ts-node` limitation that basically needs some manual setup to make things work, [more info here](https://www.npmjs.com/package/ts-node#paths-and-baseurl).

You can import your `@myrepo/config` or `@myrepo/name-of-my-library` by installing `yarn add -D tsconfig-paths` ([more info here](https://www.npmjs.com/package/tsconfig-paths)) and updating your `tsconfig.tools.json` like the following:
```diff
{
  "extends": "../tsconfig.base.json",
  "compilerOptions": {
    ...
  },
  "include": ["**/*.ts"],
+  "ts-node": {
+   "require": ["tsconfig-paths/register"]
+  }
}
```

We will update Nx to do it for you in the next versions, but for now you can do it yourself as explained.

Cheers!

**Reply by jtlapp** [reply]
Thank you! I'll use this for my next tool. I've decided that I do need an admin tool for installing/upgrading the database, and so I'll put this tool in apps. And as an side, thinking this through, I see no reason to support "down" migrations.

**Comment by jtlapp** [comment]
FYI, I am unable to build a `@nrwl/node` project that uses the postgres `pg` client, because it causes issue #13072. It seems that I'm unable to proceed with using Nx for my project.

**Reply by jtlapp** [reply]
And if I use `"executor": "@nrwl/js:tsc"`, I get "is not under 'rootDir' 'apps/installer'. 'rootDir' is expected to contain all source files" for what seems like all `index.ts` files.

**Reply by jtlapp** [reply]
I had built my libraries without the `--buildable` option, and that was causing these errors. Converting them to buildable solved the problem.

**Comment by jtlapp** [comment]
I solved this problem by (1) creating an app rather than a tool, and (2) converting my libraries to buildable libraries. I'll revisit this thread should I need to create a non-app tool. Thanks for your help!
